src/audio/smart_amp_aw/aw_smart_amp.c

```diff
@@ -0,0 +1,557 @@
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int smart_amp_ctrl_set_bin_data(struct comp_dev *dev,
+				       struct sof_ipc_ctrl_data *cdata)
+{
+	struct smart_amp_data *smartpa_data = comp_get_drvdata(dev);
+	int ret = 0;
+
+	assert(smartpa_data);
+
+	if (dev->state < COMP_STATE_READY) {
+		comp_err(dev, "[Awinic] smart_amp_ctrl_set_bin_data(): driver in init!");
+		return -EBUSY;
+	}
+
+	switch (cdata->data->type) {
+	case SOF_SMART_AMP_CONFIG:
+		ret = smart_amp_set_config(dev, cdata);
+		break;
+	case SOF_SMART_AMP_MODEL:
+		comp_err(dev, "smart_amp_ctrl_set_bin_data(): parameter type %d!", cdata->data->type);
+		break;
+	default:
+		ret = smart_amp_set_param(smartpa_data->algo_handle, dev, cdata, cdata->data->type);
+		if (ret < 0) {
+			comp_err(dev, "smart_amp_ctrl_set_bin_data(): set_param data failed!");
+		}
+		break;
+	}
+
+	return ret;
+}
+
+static int smart_amp_ctrl_set_data(struct comp_dev *dev,
+				   struct sof_ipc_ctrl_data *cdata)
+{
+	int ret = 0;
+
+	/* Check version from ABI header */
+	if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, cdata->data->abi)) {
+		comp_err(dev, "[Awinic] smart_amp_ctrl_set_data(): invalid version");
+		return -EINVAL;
+	}
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_dbg(dev, "[Awinic] smart_amp_ctrl_set_data(), SOF_CTRL_CMD_BINARY");
+		ret = smart_amp_ctrl_set_bin_data(dev, cdata);
+		break;
+	default:
+		comp_err(dev, "[Awinic] smart_amp_ctrl_set_data(): invalid cdata->cmd");
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int smart_amp_comp_cmd(struct comp_dev *dev, int cmd, void *data,
+			 int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = ASSUME_ALIGNED(data, 4);
+
+	comp_dbg(dev, "[Awinic] smart_amp_cmd(): cmd: %d", cmd);
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		return smart_amp_ctrl_set_data(dev, cdata);
+	case COMP_CMD_GET_DATA:
+		return smart_amp_ctrl_get_data(dev, cdata, max_data_size);
+	default:
+		return -EINVAL;
+	}
+}
+
+static void smart_amp_comp_free(struct comp_dev *dev)
+{
+	struct smart_amp_data *smartpa_data = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "[Awinic] smart_amp_free()");
+
+	smart_amp_deinit(smartpa_data->algo_handle, dev);
+
+	smartpa_data->algo_handle = NULL;
+	rfree(smartpa_data);
+	rfree(dev);
+}
+
+static int smart_amp_comp_params(struct comp_dev *dev,
+			    struct sof_ipc_stream_params *params)
+{
+	int err;
+
+	comp_dbg(dev, "[Awinic] smart_amp_params()");
+
+	err = comp_verify_params(dev, BUFF_PARAMS_CHANNELS, params); 
+	if (err < 0) {
+		comp_err(dev, "Awinic volume_verify_params() error: comp_verify_params() failed.");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int smart_amp_comp_trigger(struct comp_dev *dev, int cmd)
+{
+	struct smart_amp_data *smartpa_data = comp_get_drvdata(dev);
+	int ret = 0;
+
+	comp_dbg(dev, "[Awinic] smart_amp_trigger(), command = %u", cmd);
+
+	ret = comp_set_state(dev, cmd);
```

Comment:
When we allocate memory, the data is already cleared to zero, so can we not use COMP_STATE_INIT?

Reply:
sorry, not sure I understand your question. This is your `.trigger()` method, it's called for various events - start, stop, pause, etc. It can happen that an invalid state transition would be requested here, so you probably want to check that?